packages/headless/src/utils/expression-builder-sapi-v4.ts

```diff
@@ -0,0 +1,187 @@
+type QueryExpresion =
+  | KeyWord
+  | ExactMatch
+  | FieldExpresion
+  | Near
+  | Not
+  | FieldExists
+  | ObjectAccess
+  | QueryExtensionInvocation
+  | QuerySyntax
+  | And
+  | Or;
+
+type FieldValue = DateRange | Date | KeyWord | ExactMatch;
+
+type TextValue = KeyWord | ExactMatch;
+
+interface DateRange {
+  type: 'dateRange';
+  dateRange: {
+    from: string;
+    to: string;
+  };
+}
+
+interface Date {
+  type: 'date';
+  dateValue: string;
+}
+
+interface KeyWord {
+  type: 'keyword';
+  value: string;
+}
+
+interface ExactMatch {
+  type: 'exactMatch';
+  value: string;
+}
+
+interface And {
+  type: 'and';
+  expressions: QueryExpresion[];
+}
+
+interface Or {
+  type: 'or';
+  expressions: QueryExpresion[];
+}
+
+interface FieldExpresion {
+  type: 'fieldExpression';
+  fieldName: string;
+  operator:
+    | 'contains'
+    | 'isExactly'
+    | 'lowerThan'
+    | 'lowerThanOrEqual'
+    | 'greaterThan'
+    | 'greaterThanOrEqual'
+    | 'phoneticMatch'
+    | 'fuzzyMatch'
+    | 'regexMatch'
+    | 'wildcardMatch'
+    | 'differentThan';
+  value: FieldValue; // shouldn't this be an array?
+}
+
+interface Near {
+  type: 'near';
+  startTerm: TextValue;
+  otherTerms: OtherTerm[];
+}
+
+interface OtherTerm {
+  maxKeywordsFrom: number;
+  endTerm: TextValue;
+}
+
+interface Not {
+  type: 'not';
+  expression: QueryExpresion;
+}
+
+interface FieldExists {
+  type: 'fieldExists';
+  fieldName: string;
+}
+
+// Not sure how to use this one.
+interface MatchAll {
+  type: 'matchAll';
+}
+
+interface ObjectAccess {
+  type: 'objectAccess';
+  properties: string[];
+}
+
+interface QueryExtensionInvocation {
+  type: 'queryExtensionInvocation';
+  name: string;
+  argument: Argument[];
+}
+
+interface Argument {
+  name: string;
+  value: QueryExpresion;
+}
+
+interface QuerySyntax {
+  type: 'querySyntax';
+  value: string;
+}
+
+function createExpressionBuilderV4(expression?: QueryExpresion) {
```

Comment:
why is expression optional? Could we either make it obligatory or remove it, forcing users to use addExpression

Reply:
We can remove it, providing the functionality as a method where the parameter is required.
The goal is to allow developers to build on top of an existing expression. e.g.

```
const exp1 = `a AND b AND c`
const exp2 = `(exp1) OR d OR e`
```
So when building `exp2`, `exp1` would be passed, and the `build` method would concatenate the two.